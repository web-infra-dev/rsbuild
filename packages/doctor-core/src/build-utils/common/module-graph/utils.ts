import { SDK } from '@rsbuild/doctor-types';
import type { ModuleGraph } from '@rsbuild/doctor-sdk/graph';

interface DependencyData {
  type?: string;
}

/** Determine whether it is the import statement itself */
export function isImportDependency({ type }: DependencyData) {
  return (
    type === 'harmony side effect evaluation' ||
    type === 'cjs require' ||
    type === 'cjs full require' ||
    type === 'cjs export require' ||
    type === 'provided' ||
    type === 'esm import' // rspack esm import module type
  );
}

export function getImportKind(dep: DependencyData): SDK.DependencyKind {
  const { type } = dep;

  if (!type) {
    return SDK.DependencyKind.Unknown;
  }

  if (type.includes('harmony')) {
    return SDK.DependencyKind.ImportStatement;
  }

  if (type.includes('cjs')) {
    return SDK.DependencyKind.RequireCall;
  }

  if (type.includes('import()')) {
    return SDK.DependencyKind.DynamicImport;
  }

  if (type.includes('amd')) {
    return SDK.DependencyKind.AMDRequire;
  }

  return SDK.DependencyKind.Unknown;
}

/**
 * Remove the css module without connection
 *   - This will happen when the user uses the mini-css plug-in.
 */
export function removeNoImportStyle(graph: ModuleGraph) {
  graph
    .getModules()
    .filter((module) => isStyleExt(module.path))
    .filter((item) => item.getImported().length === 0)
    .forEach((item) => graph.removeModule(item));
}

export function isStyleExt(path: string) {
  return /\.(c|le|sa|sc)ss(\?.*)?$/.test(path);
}
