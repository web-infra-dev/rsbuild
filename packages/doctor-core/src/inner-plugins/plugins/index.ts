export * from './loader';
export * from './plugins';
export * from './errors';
export * from './progress';
export * from './summary';
export * from './base';
