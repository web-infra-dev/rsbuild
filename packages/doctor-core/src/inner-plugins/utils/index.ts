export * from './loader';
export * from './plugin';
export * from './sdk';
export * from './config';
export * from './circleDetect';
export * from './reportError';
