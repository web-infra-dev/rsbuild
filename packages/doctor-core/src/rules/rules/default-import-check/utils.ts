import {
  Node,
  parser,
  getDocument,
  Range,
} from '@rsbuild/doctor-utils/ruleUtils';
import { isUndefined } from 'lodash';
import type { SDK, Linter } from '@rsbuild/doctor-types';

export function getDefaultImportByRequest(node: Node.Program, request: string) {
  return parser.utils
    .getDefaultImports(node)
    .find((decl) => decl.source.value === request);
}

/**
 * Determine that it contains the same left value assignment expression in the enumeration.
 */
export function hasSameLeftInAssignStatement(
  node: Node.Program,
  compare: Node.SyntaxNode[],
) {
  return node.body
    .filter(parser.asserts.isExpressionStatement)
    .map((node) => node.expression)
    .filter(parser.asserts.isAssignmentExpression)
    .map((node) => node.left)
    .find((node) =>
      compare.some((item) => parser.utils.isSameSemantics(node, item)),
    );
}

export function importDeclarationToString(node: Node.ImportDeclaration) {
  const defaultSpecifier = node.specifiers[0] as Node.ImportDefaultSpecifier;
  const rawRequest = (node.source.raw ?? '').trim();
  return `import * as ${defaultSpecifier.local.name} from ${rawRequest}`;
}

export function getFixData(
  module: SDK.ModuleInstance,
  node: Node.ImportDeclaration,
  originalRange: Range,
): Linter.FixData {
  const { source, transformed } = module.getSource();
  const document = getDocument(source);
  const newImportText = importDeclarationToString(node);
  const oldText = transformed.substring(node.start, node.end);
  const hasLastSemi = oldText[oldText.length - 1] === ';';

  return {
    start: document.offsetAt(originalRange.start)!,
    end: document.offsetAt(originalRange.end)!,
    newText: hasLastSemi ? `${newImportText};` : newImportText,
  };
}

export function getSourceRangeFromTransformedOffset(
  module: SDK.ModuleInstance,
  node: Node.SyntaxNode,
): Linter.ReportDocument | undefined {
  const source = module.getSource();
  const sourceMap = module.getSourceMap();
  const transformedStart = node.loc!.start;
  const transformedEnd = node.loc!.end;

  if (!transformedStart || !transformedEnd) {
    return;
  }

  const transformedReport = {
    path: module.path,
    content: source.transformed,
    isTransformed: true,
    range: {
      start: transformedStart,
      end: transformedEnd,
    },
  };

  if (!sourceMap) {
    return transformedReport;
  }

  const sourceRange = module.getSourceRange(node.loc!);

  if (
    isUndefined(sourceRange?.start.line) ||
    isUndefined(sourceRange?.start.column) ||
    isUndefined(sourceRange?.end?.line) ||
    isUndefined(sourceRange?.end?.column)
  ) {
    return transformedReport;
  }

  return {
    path: module.path,
    isTransformed: false,
    content: source.source,
    range: sourceRange as Range,
  };
}
