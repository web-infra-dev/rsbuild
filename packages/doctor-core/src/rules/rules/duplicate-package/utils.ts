import path from 'path';
import bytes from 'bytes';
import type { SDK, Rule } from '@rsbuild/doctor-types';

export function getErrorMsg(
  packages: SDK.PackageInstance[],
  root: string,
): string {
  const pkgName = packages[0].name;

  let message = `Multiple packages of ${pkgName} found:\n`;

  for (const pkg of packages) {
    message += `  ${pkg.version} ${bytes(
      pkg.getSize().parsedSize,
    )} ${path.relative(root, pkg.root)}\n`;
  }

  return message.slice(0, -1);
}

export function getErrorDetail(
  pkg: SDK.PackageInstance,
  pkgGraph: SDK.PackageGraphInstance,
): Rule.PackageRelationData {
  function getImported(
    pkg: SDK.PackageInstance,
    ans: SDK.PackageDependencyInstance[],
  ): SDK.PackageDependencyInstance[] {
    const dependencies = pkgGraph.getDependenciesFromPackage(pkg);

    for (const dep of dependencies) {
      if (!dep.refDependency) {
        continue;
      }

      // The circular reference jumps out.
      if (ans.some((dep) => dep.dependency === pkg)) {
        continue;
      }

      // Go to the user code and end the query
      if (!dep.package) {
        return ans.concat(dep);
      }

      return getImported(dep.package, ans.concat(dep));
    }

    return ans;
  }

  const packageImportLinks = getImported(pkg, []);

  return {
    target: {
      name: pkg.name,
      version: pkg.version,
      root: pkg.root,
    },
    targetSize: {
      sourceSize: pkg.getSize().sourceSize,
      parsedSize: pkg.getSize().parsedSize,
    },
    dependencies: packageImportLinks.map((item) => {
      return {
        dependencyId: item.refDependency.id,
        group: item.package
          ? `${item.package.name}@${item.package.version}`
          : undefined,
      };
    }),
  };
}
