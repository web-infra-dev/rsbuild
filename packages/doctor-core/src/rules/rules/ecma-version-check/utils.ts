import type { ECMAVersion } from '@rsbuild/doctor-utils/ruleUtils';

export function getVersionNumber(ECMAString: ECMAVersion) {
  const version = ECMAString.match(/\d/);

  return version?.length ? Number(version[0]) : undefined;
}
