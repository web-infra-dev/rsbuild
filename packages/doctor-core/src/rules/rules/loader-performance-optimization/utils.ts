export function match(str: string, partterns: (string | RegExp)[]) {
  if (partterns.length === 0) return false;

  return partterns.some((p) => {
    if (typeof p === 'string') return str === p;
    if (p instanceof RegExp) return p.test(str);
    return false;
  });
}
