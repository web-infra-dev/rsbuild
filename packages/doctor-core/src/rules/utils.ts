import { Linter } from '@rsbuild/doctor-types';

export function toSeverity(
  input: Linter.SeverityInput,
  defaultLevel: Linter.Severity,
): Linter.Severity {
  if (input === 'off') {
    return Linter.Severity.Ignore;
  }

  if (input === 'on') {
    return defaultLevel;
  }

  const key = `${input[0].toUpperCase()}${input.slice(
    1,
  )}` as Linter.SeverityString;
  return Linter.Severity[key] as Linter.Severity;
}

export function noop() {}
