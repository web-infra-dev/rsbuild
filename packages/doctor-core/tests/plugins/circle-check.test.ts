import { describe, it, expect } from 'vitest';
import { checkCirclePath } from '@/inner-plugins/utils'

describe('test checkCirclePath function', () => {
  it('makeRulesSerializable()', async () => {
    const obj: Record<string, any> = { foo: { bar: [{ x: 'y' }, { y: 'x' }], xx: { yy: 'zz' } }, a: { b: 'b' } };
    obj.foo.a = obj.foo;
    obj.foo.x = obj;
    const circlePaths: string[][] = [];

    checkCirclePath(obj, [], circlePaths, 0);
    expect(circlePaths).toMatchSnapshot();

    if (circlePaths.length > 0) {
      circlePaths.forEach((_path) => {
        if (_path?.length > 0 && obj[_path[0]] !== '[Circular]') {
          obj[_path[0]] = '[Circular]';
        }
      });
    }
    JSON.stringify(obj);
    expect(obj).toMatchSnapshot();
  });
});
