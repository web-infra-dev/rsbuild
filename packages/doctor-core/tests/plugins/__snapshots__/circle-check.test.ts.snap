// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test checkCirclePath function > makeRulesSerializable() 1`] = `
[
  [
    "foo",
    "a",
  ],
  [
    "foo",
    "x",
    "foo",
  ],
]
`;

exports[`test checkCirclePath function > makeRulesSerializable() 2`] = `
{
  "a": {
    "b": "b",
  },
  "foo": "[Circular]",
}
`;
